lib/Backend/GlobOptBailOut.cpp

```diff
@@ -5,62 +5,247 @@
+GlobOpt::CaptureCopyPropValue(BasicBlock * block, Sym * sym, Value * val, SListBase<CopyPropSyms>::EditingIterator & bailOutCopySymsIter)
 {
-    if (!this->func->DoGlobOptsForGeneratorFunc())
+    if (!sym->IsStackSym())
     {
-        // TODO[generators][ianhall]: Enable constprop and copyprop for generator functions; see GlobOpt::CopyProp()
-        // Even though CopyProp is disabled for generator functions we must also not put the copy-prop sym into the
-        // bailOutInfo so that the bailOutInfo keeps track of the key sym in its byteCodeUpwardExposed list.
         return;
     }
 
-    ValueInfo * valueInfo = value->GetValueInfo();
-    Assert(stackSym->HasByteCodeRegSlot() || stackSym->HasArgSlotNum());
-    Assert(!stackSym->IsTypeSpec());
-    int32 intConstantValue;
-    if (valueInfo->TryGetIntConstantValue(&intConstantValue))
+    StackSym * copyPropSym = this->GetCopyPropSym(block, sym, val);
+    if (copyPropSym != nullptr)
     {
-        BailoutConstantValue constValue;
-        constValue.InitIntConstValue(intConstantValue);
-        bailOutInfo->capturedValues.constantValues.PrependNode(this->func->m_alloc, stackSym, constValue);
-    }
-    else if (valueInfo->IsVarConstant())
-    {
-        BailoutConstantValue constValue;
-        constValue.InitVarConstValue(valueInfo->AsVarConstant()->VarValue());
-        bailOutInfo->capturedValues.constantValues.PrependNode(this->func->m_alloc, stackSym, constValue);
-    }
-    else
-    {
-        StackSym * copyPropSym = this->GetCopyPropSym(block, stackSym, value);
-        if (copyPropSym)
-        {
-            bailOutInfo->capturedValues.copyPropSyms.PrependNode(this->func->m_alloc, stackSym, copyPropSym);
-        }
+        bailOutCopySymsIter.InsertNodeBefore(this->func->m_alloc, sym->AsStackSym(), copyPropSym);
     }
 }
 
 void
 GlobOpt::CaptureValues(BasicBlock *block, BailOutInfo * bailOutInfo)
 {
-    FOREACH_GLOBHASHTABLE_ENTRY(bucket, block->globOptData.symToValueMap)
+    if (!this->func->DoGlobOptsForGeneratorFunc())
+    {
+        // TODO[generators][ianhall]: Enable constprop and copyprop for generator functions; see GlobOpt::CopyProp()
+        // Even though CopyProp is disabled for generator functions we must also not put the copy-prop sym into the
+        // bailOutInfo so that the bailOutInfo keeps track of the key sym in its byteCodeUpwardExposed list.
+        return;
+    }
+
+    CapturedValues capturedValues;
+    SListBase<ConstantStackSymValue>::EditingIterator bailOutConstValuesIter(&capturedValues.constantValues);
+    SListBase<CopyPropSyms>::EditingIterator bailOutCopySymsIter(&capturedValues.copyPropSyms);
+
+    bailOutConstValuesIter.Next();
+    bailOutCopySymsIter.Next();
+
+    // tempBv has both changed syms with bytecode reg slot or arg slot
+    BVSparse<JitArenaAllocator> * tempBv = JitAnew(this->tempAlloc, BVSparse<JitArenaAllocator>, this->tempAlloc);
+
+    if (!block->globOptData.capturedValues)
     {
```

Comment:
Factor the code into:
if ()
{
    CaptureValueFull()
} else {
   foreach_bits() {
        ProcessCaptureVarConstValue()
        ProcessCaptureCopoySymValue()
    }
}